File: client/src/gt/component/profile/ProfileAddChar.scala
package gt.component.profile

import data.Specializations
import gt.component.GtHandler
import gt.component.widget._
import gt.component.widget.form.{GtButton, GtForm, GtInput}
import gt.service.ProfileService
import model.Toon
import org.scalajs.dom.raw.HTMLImageElement
import org.scalajs.dom.{Event, document}
import rx.{Rx, Var}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Future, Promise}
import scala.util.{Failure, Success}
import util.implicits.NodeListSeq
import util.jsannotation.js
import xuen.Component

object ProfileAddChar extends Component[ProfileAddChar](
	selector = "profile-add-char",
	templateUrl = "/assets/imports/views/profile.html",
	dependencies = Seq(GtDialog, GtBox, GtButton, GtForm, GtInput)
)

@js class ProfileAddChar extends GtHandler {
	val profile = service(ProfileService)
	val owner = attribute[Int]

	val server = Var("")
	val name = Var("")
	val spec = Var(0)
	val toon = Var(null: Toon)
	val loadDone = Var(false)
	val inFlight = Var(false)

	val specs = toon ~ {
		case null =>
			Seq.empty[Specializations.Spec]
		case t: Toon =>
			println(t.clss, Specializations.forClass(t.clss))
			Specializations.forClass(t.clss)
	}

	val hasToon = toon ~ (_ != null)
	val canLoad = Rx { server.trim.nonEmpty && name.trim.nonEmpty && !loadDone }
	val confirmDisabled = Rx { toon.! == null || inFlight }

	def loadToonImage(thumbnail: String): Future[Unit] = {
		val promise = Promise[Unit]()

		val img = document.createElement("img").asInstanceOf[HTMLImageElement]
		img.src = "https://eu.battle.net/static-render/eu/" + thumbnail.replace("avatar", "profilemain")
		child.background.appendChild(img)

		// Image loaded successfully
		img.addEventListener("load", (e: Event) => {
			img.classList.add("loaded")
			promise.success(())
		})

		// Ignore image loading failed
		img.addEventListener("error", (e: Event) => promise.success(()))

		promise.future
	}

	/** Removes old toon images */
	def removeToonImages(): Unit = {
		for (img <- child.background.querySelectorAll("img:not([default])")) {
			img.parentNode.removeChild(img)
		}
	}

	def load(): Unit = if (canLoad) {
		loadDone := true

		val input = child.as[GtInput].name
		input.error := null

		val process = for {
			available <- profile.isToonAvailable(server, name)
			_ = if (!available) throw new Exception("This character is already registered")
			toon <- profile.fetchToon(server, name)
			img <- loadToonImage(toon.thumbnail)
		} yield {
			this.toon := toon
			this.spec := toon.specid
		}

		for (e <- process.failed) {
			input.error := e.getMessage
			input.value := ""
			input.focus()
			loadDone := false
		}
	}

	def confirm(): Unit = if(!inFlight) {
		inFlight := true

		profile.registerToon(server, name, spec, owner).andThen {
			case _ => inFlight := false
		}.onComplete {
			case Failure(e) =>
				val input = child.as[GtInput].name
				input.error := e.getMessage
				input.value := ""
				loadDone := false
				toon := null
				removeToonImages()
				input.focus()

			case Success(_) =>
				child.as[GtDialog].dialog.hide()
		}
	}

	def show() = {
		child.as[GtForm].form.reset()
		Rx.flush()

		server := "sargeras"
		spec := 0
		toon := null
		loadDone := false
		inFlight := false
		Rx.flush()

		// Remove old backgrounds
		removeToonImages()

		// Show the dialog
		child.as[GtDialog].dialog.show()

		// Focus the server field
		child.name.focus()
	}
}

File: public/tools/elements/profile/ProfileAddChar.ts
import {Element, Inject, Property} from "../../polymer/Annotations";
import {GtInput} from "../widgets/GtInput";
import {GtForm} from "../widgets/GtForm";
import {GtBox} from "../widgets/GtBox";
import {PolymerElement} from "../../polymer/PolymerElement";
import {ProfileService} from "./../../services/profile/ProfileService";
import {Char} from "../../services/roster/RosterService";

///////////////////////////////////////////////////////////////////////////////
// <profile-add-char>

@Element({
	selector: "profile-add-char",
	template: "/assets/views/profile.html",
	dependencies: [GtBox, GtForm, GtInput]
})
export class ProfileAddChar extends PolymerElement {
	@Inject
	private profile: ProfileService;

	// The owner of the newly added char
	@Property
	public owner: number = this.app.user.id;

	private server: string;
	private name: string;
	private role: string;

	private char: Char;
	private load_done = false;
	private inflight = false;

	// Check if the user can load a character (required fields filled and not already loaded)
	@Property({computed: "server name load_done"})
	private get canLoad(): boolean {
		return !!this.server && !!this.name && !this.load_done;
	}

	// Load the character
	private async load() {
		if (!this.canLoad) return;
		this.load_done = true;

		let input: GtInput = this.$.name;
		input.error = null;

		try {
			// Check if the character is already registered to someone in the database
			let available = await this.profile.checkAvailability(this.server, this.name);
			if (!available) {
				throw new Error("This character is already registered");
			}

			// Fetch the char from Battle.net
			let char = await this.profile.fetchChar(this.server, this.name);

			// Change the dialog background
			let img = document.createElement("img");
			let background: HTMLDivElement = this.$.background;
			img.src = "http://eu.battle.net/static-render/eu/" + char.thumbnail.replace("avatar", "profilemain");
			Polymer.dom(background).appendChild(img);
			//img.onload = () => {
				this.char = char;
				this.role = char.role;
				img.classList.add("loaded");
			//};
		} catch (e) {
			input.error = e.message;
			input.value = "";
			input.focus();
			this.load_done = false;
		}
	}

	// Role selected
	private roleClicked(e: MouseEvent) {
		let img = <HTMLImageElement> e.target;
		this.role = img.getAttribute("role");
	}

	// Show the add-char dialog
	public show() {
		// Reset fields values
		this.name = "";
		this.server = "";
		this.char = null;
		this.load_done = false;

		// Remove old backgrounds
		let background: HTMLDivElement = this.$.background;
		let imgs = background.querySelectorAll<HTMLImageElement>("img:not([default])");
		for (let i = 0; i < imgs.length; i++) {
			imgs[i].remove();
		}

		// Show the dialog
		this.$.dialog.show();

		// Focus the server field
		this.$.name.focus();
	}

	@Property({computed: "char inflight"})
	private get confirmDisabled(): boolean {
		return !this.char || this.inflight;
	}

	// Add the character to the user account
	private async confirm() {
		let server = this.$.server.value;
		let name = this.$.name.value;
		let role = this.role;
		let owner = this.owner;

		this.inflight = true;
		try {
			await this.profile.registerChar(server, name, role, owner);
			this.close();
		} catch (e) {
			let input: GtInput = this.$.name;
			input.error = e.message;
			input.value = "";
			input.focus();
			this.load_done = false;
		} finally {
			this.inflight = false;
		}
	}

	// Close the dialog
	public close() {
		this.$.dialog.hide();
	}
}
